pyop2/global_kernel.py

```diff
@@ -405,20 +405,26 @@ def _ldargs(self):
         return tuple(ldargs)
 
 
-@serial_cache(hashkey=lambda knl: knl.cache_key)
-def _generate_code_from_global_kernel(kernel):
-    with PETSc.Log.Event("GlobalKernel: generate loopy"):
-        wrapper = generate(kernel.builder)
-
-    with PETSc.Log.Event("GlobalKernel: generate device code"):
-        code = lp.generate_code_v2(wrapper)
-
-    if kernel.local_kernel.cpp:
-        preamble = "".join(process_preambles(getattr(code, "device_preambles", [])))
-        device_code = "\n\n".join(str(dp.ast) for dp in code.device_programs)
-        return preamble + "\nextern \"C\" {\n" + device_code + "\n}\n"
+@memory_and_disk_cache(hashkey=lambda knl, _: knl.cache_key)
+@mpi.collective
+def _generate_code_from_global_kernel(kernel, comm):
+    if comm.rank == 0:
+        with PETSc.Log.Event("GlobalKernel: generate loopy"):
+            wrapper = generate(kernel.builder)
+
+        with PETSc.Log.Event("GlobalKernel: generate device code"):
+            code = lp.generate_code_v2(wrapper)
+
+        if kernel.local_kernel.cpp:
+            preamble = "".join(process_preambles(getattr(code, "device_preambles", [])))
+            device_code = "\n\n".join(str(dp.ast) for dp in code.device_programs)
+            code = preamble + "\nextern \"C\" {\n" + device_code + "\n}\n"
+        else:
+            code = code.device_code()
+    else:
+        code = None
 
-    return code.device_code()
+    return comm.bcast(code, root=0)
```

Comment:
Can we do `bcast` in `compile_global_kernel`; this way, `_generate_code_from_global_kernel` can remain unaware of `comm`.

Reply:
Unsure. I want:
* The resulting code to be cached on disk
* The code generation to be done on only one rank (Pablo gave some evidence that doing otherwise was slow)
* Each rank to have the generated code kept in memory

Reply:
I think you are right. We want to do it this way.

Reply:
I have now added a `bcast` kwarg to `parallel_cache` (defaults to `False`) so we can get this behaviour more transparently.